rtengine/dcraw.cc

```diff
@@ -2417,59 +2417,78 @@ void CLASS hasselblad_correct()
 
 void CLASS hasselblad_load_raw()
 {
-  struct jhead jh;
-  int shot, row, col, *back[5], len[2], diff[12], pred, sh, f, s, c;
-  unsigned upix, urow, ucol;
-  ushort *ip;
+    struct jhead jh;
+    int *back[5], diff[12];
 
-  if (!ljpeg_start (&jh, 0)) return;
-  order = 0x4949;
-  ph1_bithuff_t ph1_bithuff(this, ifp, order);
-  hb_bits(-1);
-  back[4] = (int *) calloc (raw_width, 3*sizeof **back);
-  merror (back[4], "hasselblad_load_raw()");
-  FORC3 back[c] = back[4] + c*raw_width;
-  cblack[6] >>= sh = tiff_samples > 1;
-  shot = LIM(shot_select, 1, tiff_samples) - 1;
-  for (row=0; row < raw_height; row++) {
-    FORC4 back[(c+3) & 3] = back[c];
-    for (col=0; col < raw_width; col+=2) {
-      for (s=0; s < tiff_samples*2; s+=2) {
-	FORC(2) len[c] = ph1_huff(jh.huff[0]);
-	FORC(2) {
-	  diff[s+c] = hb_bits(len[c]);
-	  if ((diff[s+c] & (1 << (len[c]-1))) == 0)
-	    diff[s+c] -= (1 << len[c]) - 1;
-	  if (diff[s+c] == 65535) diff[s+c] = -32768;
-	}
-      }
-      for (s=col; s < col+2; s++) {
-	pred = 0x8000 + load_flags;
-	if (col) pred = back[2][s-2];
-	if (col && row > 1) switch (jh.psv) {
-	  case 11: pred += back[0][s]/2 - back[0][s-2]/2;  break;
-	}
-	f = (row & 1)*3 ^ ((col+s) & 1);
-	FORC (tiff_samples) {
-	  pred += diff[(s & 1)*tiff_samples+c];
-	  upix = pred >> sh & 0xffff;
-	  if (raw_image && c == shot)
-	    RAW(row,s) = upix;
-	  if (image) {
-	    urow = row-top_margin  + (c & 1);
-	    ucol = col-left_margin - ((c >> 1) & 1);
-	    ip = &image[urow*width+ucol][f];
-	    if (urow < height && ucol < width)
-	      *ip = c < 4 ? upix : (*ip + upix) >> 1;
-	  }
-	}
-	back[2][s] = pred;
-      }
+    if (!ljpeg_start (&jh, 0)) {
+        return;
+    }
+    order = 0x4949;
+    ph1_bithuff_t ph1_bithuff(this, ifp, order);
+    hb_bits(-1);
+    back[4] = (int *) calloc(raw_width, 3 * sizeof **back);
+    merror(back[4], "hasselblad_load_raw()");
+    for (int c = 0; c < 3; ++c) {
+        back[c] = back[4] + c * raw_width;
+    }
+    const int sh = tiff_samples > 1;
+    cblack[6] >>= sh;
+    const int shot = LIM(shot_select, 1, tiff_samples) - 1;
+    for (int row = 0; row < raw_height; ++row) {
+        for (int c = 0; c < 4; ++c) {
+            back[(c + 3) & 3] = back[c];
+        }
+        for (int col = 0; col < raw_width; col += 2) {
+            for (int s = 0; s < tiff_samples * 2; s += 2) {
+                int len[2];
+	            for (int c = 0; c < 2; ++c) {
```

Comment:
ph1_huff() reads from file changes filepos. I will change it to the const :+1:

Reply:
@Floessie Is it guaranteed that the first ph1_huff call in the assignment is executed before the second?

Reply:
I guess so. From [here](https://en.cppreference.com/w/cpp/language/list_initialization):

> Every initializer clause is sequenced before any initializer clause that follows it in the braced-init-list. This is in contrast with the arguments of a function call expression, which are unsequenced.

And from [here](https://en.cppreference.com/w/c/language/array_initialization):

> The order of evaluation of subexpressions in an array initializer is indeterminately sequenced in C (but not in C++ since c++11)

HTH,
Flössie

Reply:
Thank you :+1: Fixed